File: packages/core/src/programmers/iterate/check_native.ts
import ts from "@typescript/native-preview";

import { ExpressionFactory } from "../../factories/ExpressionFactory";

/** @internal */
export const check_native = (props: {
  name: string;
  input: ts.Expression;
}): ts.Expression => {
  const instanceOf = ExpressionFactory.isInstanceOf(props.name, props.input);
  return ATOMIC_LIKE.has(props.name)
    ? ts.factory.createLogicalOr(
        ts.factory.createStrictEquality(
          ts.factory.createStringLiteral(props.name.toLowerCase()),
          ts.factory.createTypeOfExpression(props.input),
        ),
        instanceOf,
      )
    : instanceOf;
};

/** @internal */
const ATOMIC_LIKE = new Set(["Boolean", "Number", "String"]);

File: packages/core/src/programmers/iterate/check_native.go
package iterate

import h "github.com/samchon/typia/packages/core/src/programmers/helpers"

func check_native(input h.Expression, name string) h.Expression {
	switch name {
	case "Boolean", "Number", "String":
		return h.Expr(string(input) + " instanceof " + name + " || typeof " + string(input) + " === " + h.QuoteString(lowerNative(name)))
	case "Array":
		return h.Expr("Array.isArray(" + string(input) + ")")
	default:
		return h.Expr(string(input) + " instanceof " + name)
	}
}

func lowerNative(name string) string {
	switch name {
	case "Boolean":
		return "boolean"
	case "Number":
		return "number"
	case "String":
		return "string"
	default:
		return name
	}
}
